### PR TITLE
Make ju.Objects.requireNonNull overloads follow Semantics for UB NPE.

### PR DESCRIPTION
**Pull request by sjrd**
And use it for all explicit null checks in the javalib.

Unfortunately, it is unclear what to do about the two overloads of `ju.Objects.requireNonNull` with an explicit message. Either we keep the message and we have to throw an explicit NPE; or we trigger a UB NPE, but we ignore the message.

We choose to trigger the UB, but then we catch a genuine NPE to rethrow it with the correct message. The message will still be lost when the NPEs are Fatal, but at least they are used when NPEs are Compliant.

For the overload that takes a `messageSupplier`, additionally we choose to only call its `get()` method when we actually need the message, i.e., only in Compliant mode.

These changes allow for good optimization opportunities. We intrinsify the 3 overloads so we can get the best possible code for every checked behavior. For Unchecked mode, the tests completely disappear. For Compliant and Fatal, we don't need to throw a fake NPE that we immediately try to catch and re-throw.

---

Extracted the NPE-related stuff from #5329.

### CODE REVIEW

File: javalib/src/main/scala/java/lang/_String.scala [hunk 1]
```diff
@@ -348,22 +346,24 @@ final class _String private () // scalastyle:ignore
...
-      prefix.getClass() // null check
-      (toffset <= length() && toffset >= 0 &&
-        thisString.asInstanceOf[js.Dynamic].startsWith(prefix, toffset).asInstanceOf[scala.Boolean])
-    } else {
-      (toffset <= length() && toffset >= 0 &&
-      thisString.jsSubstring(toffset, toffset + prefix.length()) == prefix)
+    toffset <= length() && toffset >= 0 && {
+      if (LinkingInfo.esVersion >= ESVersion.ES2015) {
+        thisString.asInstanceOf[js.Dynamic]
+          .startsWith(requireNonNull(prefix), toffset)
```
Comment (gzm0) [inline javalib/src/main/scala/java/lang/_String.scala:390]: Doesn't this swallow NPEs in case the offset is out of bounds? (if yes, fix tests?)

**Comment by gzm0** [comment]
I think this comment still applies / is unaddressed: https://github.com/scala-js/scala-js/pull/5329#discussion_r2936617822

**Review by gzm0 — CHANGES_REQUESTED** [review]
Review of everything but the first commit.

File: javalib/src/main/scala/java/lang/_String.scala (hunk 1 above)
Comment (sjrd) [reply javalib/src/main/scala/java/lang/_String.scala:390]: Huh, yes, it does, and it's actually unclear which behavior is correct. I have extracted this particular change in a separate commit, with accompanying tests and comments.

**Comment by sjrd** [comment]
Continuing the discussion from https://github.com/scala-js/scala-js/pull/5329#discussion_r2936538615

> > If we observe a genuine NPE without optimizer, then we must also observe a genuine NPE with optimizer.
> 
> It is not clear to me what you mean by "genuine" here. Are you saying that if we observe an NPE under non-compliant NPE semantics w/o optimizer we also have to observe one with the optimizer?
> 
> If yes, that seems to be very restrictive. I would have expected not having to do that to precisely be the point of UB 🤔

By genuine NPE I mean that we observe a catchable `NullPointerException` (not wrapped in an `UndefinedBehaviorError`).

Our UB system is built on the following principle: if, in Fatal mode, I do not observe an `UndefinedBehaviorError`, then my code will behave exactly the same in Unchecked mode. (1) That is important so that users can test in Fatal (fastLink) mode, and get guarantees about their Unchecked (fullLink) mode.

Our optimizer is built on the following principle: if the code does not run into an Unchecked UB without the optimizer, then it behaves the same with the optimizer. (2) i.e., the optimizer does not alter non-Unchecked UB semantics (modulo details of error messages).

From there, we can follow logic steps with your proposal:

1. Without optimizer and with Fatal mode, I observe an NPE (because that's in the javalib code).
2. Therefore with optimizer and with Fatal mode, by (2) I must observe an NPE.
3. Therefore with optimizer and with Unchecked mode, by (1) I must observe an NPE.

By taking the contrapositions, if we want to optimize away the null check in Unchecked mode with optimizer, we must not observe an NPE without optimizer and in Fatal mode. Therefore, the javalib code must not throw an NPE / be written as if Compliant.

**Comment by gzm0** [comment]
> Our UB system is built on the following principle: if, in Fatal mode, I do not observe an UndefinedBehaviorError, then my code will behave exactly the same in Unchecked mode. (1) That is important so that users can test in Fatal (fastLink) mode, and get guarantees about their Unchecked (fullLink) mode.

Right... That's for course quite restrictive for us, but very important for the users.

For the future: I have also thought about the idea of intercepting exceptions that result in UBEs entirely in the backend:

```scala
// In the backend re-write this:
throw new NPE
// to
throw new UBE(new NPE)
// of course, this breaks as soon as we have
val x = new NPE
throw x
// lest we do runtime type checking of the exception, which we definitely do not.
```

So all in all, I'm convinced that the solution in this PR is the best option we have. I'll proceed to in-detail review.

File: javalib/src/main/scala/java/util/Objects.scala [hunk 2]
```diff
@@ -84,8 +87,61 @@ object Objects {
   def nonNull(obj: Any): Boolean =
     obj != null
 
+  // Intrinsic
   @inline
   def requireNonNull[T](obj: T, messageSupplier: Supplier[String]): T =
-    if (obj == null) throw new NullPointerException(messageSupplier.get())
+    if (obj == null) throwNPEWithMessage(messageSupplier)
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Objects.scala:93]: I do not understand why we need two overloads of `throwNEPWithMessage`.

Can't we call `messageSupplier.get()` here?

- if it is `null`, it will trigger NPE with unspecified message right away
- otherwise we get the message and then do the throw spiel (IIUC this will result in different observable behavior with and without optimizer under fatal NPEs, but IIUC that is OK).

File: javalib/src/main/scala/java/util/Objects.scala [hunk 3]
```diff
@@ -84,8 +87,61 @@ object Objects {
...
+   * semantics if its `get()` method threw an exception itself.
+   *
+   * ---
+   *
+   * The methods are `@noinline` because they are in a slow path anyway.
+   * We don't want the try..catch'es to appear at call site. That pollutes
+   * performance for the entire enclosing function in some engines.
+   */
+
+  @noinline private def throwNPEWithMessage(message: String): Nothing = {
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Objects.scala:130]: Have you considered intrinsifying this as well and rewrite to `throw new NullPointerException(message)` in compliant mode?

IIUC that would allow you to re-use the bodies of the call sites in their intrinsics.

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala [hunk 4]
```diff
@@ -3341,6 +3341,78 @@ private[optimizer] abstract class OptimizerCore(
...
+
+              cont1(PreTransTree(Block(
+                If(BinaryOp(BinaryOp.===, objLocalDef.newReplacement, Null()), {
+                  UnaryOp(UnaryOp.Throw,
+                      New(NullPointerExceptionClass,
+                          MethodIdent(StringArgConstructorName), List(actualMessage)))
+                }, {
+                  finishTransformExpr(foldCast(objLocalDef.toPreTransform, resultType))
+                })(resultType)
+              )))
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:3395]: This looks fine but I believe it can be simplified if we intrinsify `throwNPEWithMessage`

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ObjectsTest.scala [hunk 5]
```diff
@@ -87,11 +91,67 @@ class ObjectsTest {
...
+
+      val e2 = assertThrows(classOf[NullPointerException],
+          ju.Objects.requireNonNull(hide[String](null), "the message"))
+      assertEquals("the message", e2.getMessage())
+    }
+
+    assertEquals("abc", ju.Objects.requireNonNull("abc", "unexpected"))
+    assertEquals("abc", ju.Objects.requireNonNull(hide[String]("abc"), "unexpected"))
+
+    // The effects of computing the arguments are preserved, in order
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ObjectsTest.scala:121]: IIUC these test cases are only necessary because we cannot rely on the inliner in the Optimizer, right?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ObjectsTest.scala [hunk 6]
```diff
@@ -108,11 +168,70 @@ class ObjectsTest {
...
+      assertNull(e3.getMessage())
+      val e4 = assertThrows(classOf[NullPointerException],
+          ju.Objects.requireNonNull(hide[String](null), () => null))
+      assertNull(e4.getMessage())
+
+      // If the supplier itself is null as well, we get an NPE with an unspecified message
+      assertThrows(classOf[NullPointerException],
+          ju.Objects.requireNonNull(null, null: ju.function.Supplier[String]))
+      assertThrows(classOf[NullPointerException],
+          ju.Objects.requireNonNull(hide[String](null), null: ju.function.Supplier[String]))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ObjectsTest.scala:203]: Do we need variants where we hide the null supplier as well?

**Review by gzm0 — CHANGES_REQUESTED** [review]

File: javalib/src/main/scala/java/util/Objects.scala (hunk 2 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Objects.scala:93]: We could choose to call `messageSupplier.get()` here. The following paragraph of the commit message calls attention to the choice we made instead:

> For the overload that takes a `messageSupplier`, additionally we choose to only call its `get()` method when we actually need the message, i.e., only in Compliant mode.

Calling `messageSupplier.get()` here would not allow to remove the checks in Unchecked. If we call `.get()` here *and* we remove the check in Unchecked, we have cases where design principle (1) is violated:
```scala
requireNonNull(null, () => throw new IllegalArgumentException())
```
* in Fatal mode, this throws an `IllegalArgumentException`, because we performed `messageSupplier.get()` immediately after testing `obj == null`.
* in Unchecked mode, it doesn't throw.

File: javalib/src/main/scala/java/util/Objects.scala (hunk 3 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Objects.scala:130]: Whenever possible, we try to only intrinsify publicly specified methods, rather than private implementation details. This makes sure that (past and) future implementation changes to the javalib do not affect what gets intrinsified. So far, the only non-public methods we intrinsify are the two methods in `scala.c.m.GenericArrayBuilder`, and I'm already not proud of those.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ObjectsTest.scala (hunk 5 above)
Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ObjectsTest.scala:121]: We need that kind of tests for intrinsincs. Because if we mess up the replacement code for the intrinsic, we may alter evaluation order compared to the regular behavior.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ObjectsTest.scala (hunk 6 above)
Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ObjectsTest.scala:203]: Huh. Yes, we definitely that. It uncovered a bug 😱  Nice catch!

File: javalib/src/main/scala/java/util/Objects.scala (hunk 2 above)
Comment (gzm0) [reply javalib/src/main/scala/java/util/Objects.scala:93]: 😱 right :-/ I'll suggest a comment at the call site of `get()`.

File: javalib/src/main/scala/java/util/Objects.scala [hunk 7]
```diff
@@ -84,8 +87,61 @@ object Objects {
...
+        throw new NullPointerException(message)
+    }
+  }
+
+  @noinline private def throwNPEWithMessage(messageSupplier: Supplier[String]): Nothing = {
+    try {
+      throw null
+    } catch {
+      case _: NullPointerException =>
+        throw new NullPointerException(messageSupplier.get())
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Objects.scala:148]:
```suggestion
        // It is important that we only call `messageSupplier.get()` here (after UB)
        // Otherwise a throwing messageSupplier might mask UB in unchecked.
        // For example in: `requireNonNull(null, () => throw new IllegalArgumentException())`.
        throw new NullPointerException(messageSupplier.get())
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ObjectsTest.scala [hunk 8]
```diff
@@ -101,18 +161,93 @@ class ObjectsTest {
...
+      assertThrows(classOf[NullPointerException],
+          ju.Objects.requireNonNull(null, hide[ju.function.Supplier[String]](null)))
+      assertThrows(classOf[NullPointerException],
+          ju.Objects.requireNonNull(hide[String](null), hide[ju.function.Supplier[String]](null)))
+    }
 
     assertEquals("abc", ju.Objects.requireNonNull("abc", failureSupplier))
+    assertEquals("abc", ju.Objects.requireNonNull(hide[String]("abc"), failureSupplier))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ObjectsTest.scala:211]: Also add tests with `null` suppliers?

**Review by gzm0 — APPROVED** [review]
LGTM modulo some minor suggestions.

**Comment by sjrd** [comment]
Rebased to fix a trivial conflict in `import`s, and added the two suggestions.
